### DISCUSSION TITLE
Wait for the canvas to be ready (after a SetSettings call) ?

### DISCUSSION THREAD

**Discussion by 0gust1** [molstar/molstar#473] [Q&A] [answered]
We're writing a Svelte wrapper for Molstar, but I guess the question is not Svelte related.

During the initialization phase, we initialize the Canvas with `PluginCommands.Canvas3D.SetSettings()`

```
PluginCommands.Canvas3D.SetSettings(plugin, {
			settings: {
				renderer: {
					...renderer,
					ambientIntensity: 1.0,
					light: [],
					backgroundColor: 0x001f42 as Color
				},
				postprocessing: {
					occlusion: {
						name: 'on',
						params: { samples: 32, radius: 6, bias: 1.4, blurKernelSize: 15 }
					},
					outline: { name: 'on', params: { scale: 1, threshold: 0.33, color: Color(0x000000) } }
				}
			}
		});
```

and then, we load PDB structures.

When loaded, the structures render as the screenshot (notice the absent outlines).

![Capture d’écran 2022-07-04 à 12 58 43](https://user-images.githubusercontent.com/578154/177141394-a4f96970-f9e9-4965-b86c-e8360cd55b3a.png)

If we programmatically call again `PluginCommands.Canvas3D.SetSettings()` **after** the structures are loaded, we get the correct rendering.

![Capture d’écran 2022-07-04 à 13 03 15](https://user-images.githubusercontent.com/578154/177142196-753421c0-2e9f-4db6-85cd-e662ae257f69.png)

## Question(s):

It seems the renderer/postprocessing are not completely initialized when we load the structure. We tried to `await` the call to `PluginCommands.Canvas3D.SetSettings()` but no luck.

Is there something we can call or an event/subscribable that we can use to be sure that the Canvas3D is ready before adding structures to the scene ?

**Comment by 0gust1** [comment, marked as answer]
I just updated my project to use `molstar@latest`  (`3.10.2`)

without changing anything on my code, we have now the following result (_edit: fixed, see below_)
![Capture d’écran 2022-07-04 à 15 40 18](https://user-images.githubusercontent.com/578154/177167033-fa37c2c2-8d20-4c74-9c52-4079d447a293.png)

- ~we lost occlusion shadows and ribbon shading~ (fixed: see below)
- we lost some things: water molecules and some things
- **we don't have no more the discrepancy between the initial render and manual call to `PluginCommands.Canvas3D.SetSettings()`**

We may eventually consider that the problem stated is no more here, but we still would like to get the occlusion shadows & ribbon shading back (it helps a lot the readability).

I'll update the conversation if we manage to get all those things back.

**(edit 1):**

we got the occlusion and shading back (`resolutionScale: 1` added in `postprocessing.occlusion.params`):
```
await PluginCommands.Canvas3D.SetSettings(plugin, {
			settings: {
				renderer: {
					...renderer,
					ambientIntensity: 1.0,
					light: [],
					backgroundColor: 0x001f42 as Color
				},
				postprocessing: {
					occlusion: {
						name: 'on',
						params: { samples: 32, radius: 6, bias: 1.4, blurKernelSize: 15, resolutionScale: 1}
					},
					outline: { name: 'on', params: { scale: 1, threshold: 0.33, color: Color(0x000000) } }
				}
			}
		});
```

**Reply by dsehnal** [reply]
Glad  you could find a fix :)
